clients/client-common/src/main/java/org/eclipse/hono/client/util/CachingClientFactory.java

```diff
@@ -63,9 +59,23 @@
      */
     private final Map<String, T> activeClients = new HashMap<>();
     /**
-     * The locks for guarding the creation of new instances.
+     * The locks for guarding the creation of new instances. Map key is the client cache key.
      */
     private final Map<String, Boolean> creationLocks = new HashMap<>();
+    /**
+     * List of request data for client creation requests that are put on hold because
+     * a concurrent request is still not completed.
+     */
+    private final List<CreationRequestData> waitingCreationRequests = new LinkedList<>();
```

Comment:
wouldn't access to the requests be more efficient if we used a Map<String, Deque<CreationRequestData>> instead? We then wouldn't need to iterate over all elements every time a request for a particular key fails/succeeds, would we?

Reply:
I've deliberately chosen a data structure to keep the request order across requests with different keys.

As an example, let's say there are client creation requests for `keyA`, 2 times `keyB`, and then 100 times `keyA`.
When creation for `keyA` succeeds, the first 10 ( `WAITING_CREATION_REQUESTS_COMPLETION_BATCH_SIZE_DEFAULT`) of the 100 waiting `keyA` requests will get handled.
When in the mean time creation of `keyB` succeeded, the decoupled `processWaitingCreationRequests` invocation shall then handle the 2nd `keyB` request first, before the 90 `keyA` requests.

Reply:
I see